### PR TITLE
re-scope scoped attributes when an attribute with the same name and d…

### PR DESCRIPTION
**Pull request by knoxsp**
…imension at a higher scope is added

### CODE REVIEW

File: hydra_base/lib/attributes.py [hunk 1]
```diff
@@ -195,14 +226,78 @@ def _add_attribute(attr, user_id, flush=True):
...
+    #Reassign all resource attributes which point to scoped attirbutes, and then delete
+    #the scoped attributes.
+    scoped_resource_attrs_qry = db.DBSession.query(ResourceAttr).join(Attr).filter(
+        ResourceAttr.attr_id == Attr.id,
+        func.lower(Attr.name) == attr_i.name.lower(),
+        Attr.dimension_id == attr_i.dimension_id,
+        Attr.id != attr_id
+    )
+
+    log.info("%s scoped attribuytes found with same name & dimension. Reassigning.")
```
Comment (pmslavin) [inline hydra_base/lib/attributes.py:279]: For future greps...
```suggestion
    log.info("%s scoped attributes found with same name & dimension. Reassigning.")
```

**Review by pmslavin — COMMENTED** [review]
Tests all pass for me...
```
attributes/test_attributes.py .................................................                                                                                                                   
attributes/test_attributes_scoped.py ...........
```

Comments are mainly changing multi-line comments to strings.

One question: What happens if `_reassign_scoped_attributes(attr_id)` is called with a network-scoped attr_id, and the `matching_attrs` (L259) includes attrs with a *higher* scope level?

**Comment by knoxsp** [comment]
To answer your question, this should never occur, because the function call is in an if statement, checking to make sure the attr is not network scoped.

**Comment by pmslavin** [comment]
Tests all pass...
```
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.10, pytest-7.1.2, pluggy-0.13.1
db-backend: mysql
connection-type: local
rootdir: /home/paul/code/Hydra/hydra-base
plugins: order-1.0.1, cov-3.0.0
collected 328 items                                                                                                                                                                          

tests/test_data.py .............................                                                                                                                                       [  8%]
tests/test_datatypes.py .......................................                                                                                                                        [ 20%]
tests/test_groups.py .....                                                                                                                                                             [ 22%]
tests/test_mongo.py .......                                                                                                                                                            [ 24%]
tests/test_network.py ...............................                                                                                                                                  [ 33%]
tests/test_rules.py ..................                                                                                                                                                 [ 39%]
tests/test_scenario.py ................................                                                                                                                                [ 49%]
tests/test_units.py .....................                                                                                                                                              [ 55%]
tests/test_users.py ..................                                                                                                                                                 [ 60%]
tests/attributes/test_attributes.py .................................................                                                                                                  [ 75%]
tests/attributes/test_attributes_scoped.py ..........                                                                                                                                  [ 78%]
tests/project/test_project.py ...........                                                                                                                                              [ 82%]
tests/project/test_project_inheritance.py .......                                                                                                                                      [ 84%]
tests/templates/test_template_inheritance.py ...........                                                                                                                               [ 87%]
tests/templates/test_templates.py ..................................                                                                                                                   [ 98%]
tests/test_login.py ......                                                                                                                                                             [100%]

```

File: hydra_base/lib/attributes.py [hunk 2]
```diff
@@ -138,44 +124,280 @@ def get_template_attributes(template_id, **kwargs):
...
     except NoResultFound:
         return None
 
-@required_role('admin')
-def add_attribute_no_checks(attr,**kwargs):
+def search_attributes(name, network_id=None, project_id=None, **kwargs):
     """
-    ***WARNING*** This is used for test purposes only, and can allow
-    duplicate attributes to be created
+        Search for all attributes matching the given name
```
Comment (pmslavin) [inline hydra_base/lib/attributes.py:171]:
```suggestion
        Search for all attributes matching the given name, ignoring case
```

**Review by pmslavin — APPROVED** [review]
All tests pass, tested with HWI's dataset-comparison and scoped-attr branches.
Any comments are minor and do not affect functionality so is approved.

**Pull request merged**